addons/project/models/project_task_type.py

```diff
@@ -79,6 +84,64 @@ def copy(self, default=None):
+            [('user_id', 'in', personal_stages.user_id.ids), ('id', 'not in', personal_stages.ids)],
+            groupby=['user_id', 'sequence', 'id'],
+            order="user_id,sequence DESC",
+        )
+        remaining_personal_stages_by_user = defaultdict(list)
+        for user, sequence, stage in remaining_personal_stages_all:
+            remaining_personal_stages_by_user[user].append({'id': stage.id, 'seq': sequence})
+
+        # For performance issue, project.task.stage.personal records that need to be modified are listed before calling _prepare_personal_stages_deletion
+        personal_stages_to_update = self.env['project.task.stage.personal']._read_group([('stage_id', 'in', personal_stages.ids)], ['stage_id'], ['id:recordset'])
+        for user in personal_stages.user_id:
+            if not user.active or not user.has_group('base.group_user'):
+                continue
+            user_stages_to_unlink = personal_stages.filtered(lambda stage: stage.user_id == user)
+            user_remaining_stages = remaining_personal_stages_by_user[user]
+            if not user_remaining_stages:
+                raise UserError(_("Each user should have at least one personal stage. Create a new stage to which the tasks can be transferred after the selected ones are deleted."))
+            self._prepare_personal_stages_deletion(user_stages_to_unlink, user_remaining_stages, personal_stages_to_update)
+
+        # 2. Deleting all stages in self (personal and project stages)
+        return super().unlink()
+
+    def _prepare_personal_stages_deletion(self, stages_to_delete, remaining_stages_dict, personal_stages_to_update):
+        """ _prepare_personal_stages_deletion prepare the deletion of personal stages of a single user.
+            Tasks using that stage will be moved to the first stage with a lower sequence if it exists
+            higher if not.
+        :param stages_to_delete: project.task.type recordset containing the personal stage of a user
+                                 that need to be deleted
+        :param remaining_stages_dict: list of dict representation of the personal stages of a user that
+                                      can be used to replace the deleted ones. Can not be empty.
+                                      e.g: [{'id': stage1_id, 'seq': stage1_sequence}, ...]
+        :param personal_stages_to_update: project.task.stage.personal recordset containing the records
+                                          that need to be updated after stage modification. Is passed to
+                                          this method as an argument to avoid to reload it for each users
+                                          when this method is called multiple times.
+        """
+        stages_to_delete_dict = sorted(stages_to_delete.mapped(lambda s: {'id': s.id, 'seq': s.sequence}),
+                                       key=lambda stage: stage['seq'])
```

Comment:
Cosmetic: Seems like a disguised dict-comprehension.
```suggestion
        stages_to_delete_dict = sorted(
            {'id': s.id, 'seq': s.sequence for s in stages_to_delete}),
            key=lambda stage: stage['seq'])
```

Reply:
> Cosmetic: Seems like a disguised dict-comprehension.

Indeed :+1: But more a disguised list (of dict) comprehension in that case I would say:
```suggestion
        stages_to_delete_dict = sorted([{'id': stage.id, 'seq': stage.sequence} for stage in stages_to_delete],
                                       key=lambda stage: stage['seq'])

```
(still have to test it)